lib/supavisor/client_handler.ex

```diff
@@ -352,32 +374,60 @@ defmodule Supavisor.ClientHandler do
     :undef
   end
 
-  @spec handle_exchange(sock(), fun(), boolean()) :: :ok | {:error, String.t()}
-  def handle_exchange({_, socket} = sock, password, ssl) do
+  @spec handle_exchange(sock(), {atom(), fun()}, boolean()) ::
+          {:ok, binary() | nil} | {:error, String.t()}
+  def handle_exchange({_, socket} = sock, {method, secrets}, ssl) do
```

Comment:
This function now is very quite long and nested, it may be worth refactoring it. One option is to use `with` to handle the common error clauses. Here is an attempt:

```elixir
defp receive_next(socket, timeout_message) do
  receive do
    {_proto, ^socket, bin} ->
      Server.decode_pkt(bin)
  after
    15_000 -> {:error, timeout_message}
  end
end

defp reply_first_exchange(sock, method, secrets, nonce, user, ssl) do
  channel = if ssl, do: "eSws", else: "biws"
  {message, signatures} = exchange_first(method, secrets, nonce, user, channel)
  :ok = sock_send(sock, Server.exchange_message(:first, message))
end

defp authenticate_exchange(:password, secrets, p) do
  if p == signatures.client do
    {:ok, nil}
  else
    {:error, "Invalid client signature"}
  end
end

defp authenticate_exchange(:auth_query, secrets, p) do
  client_key = :crypto.exor(p |> Base.decode64!(), signatures.client)

  if H.hash(client_key) == secrets.().stored_key do
    {:ok, client_key}
  else
    {:error, "Invalid client signature"}
  end
end

@spec handle_exchange(sock(), {atom(), fun()}, boolean()) ::
        {:ok, binary() | nil} | {:error, String.t()}
def handle_exchange({_, socket} = sock, {method, secrets}, ssl) do
  :ok = sock_send(sock, Server.auth_request())

  with {:ok,
        %{
          tag: :password_message,
          payload: {:scram_sha_256, %{"n" => user, "r" => nonce}}
        }, _} <- receive_next(socket, "Timeout while waiting for the first password message"),
       reply_first_exchange(sock, method, secrets, nonce, user, ssl),
       {:ok,
        %{
          tag: :password_message,
          payload: {:first_msg_response, %{"p" => p}}
        }, _} <- receive_next(socket, "Timeout while waiting for the second password message"),
       {:ok, key} <- authenticate_exchange(socket, method, secrets, signatures, p) do
    message = "v=#{Base.encode64(signatures.server)}"
    :ok = sock_send(sock, Server.exchange_message(:final, message))
    {:ok, key}
  else
    {:error, message} -> {:error, message}
    other -> {:error, "Unexpected message #{inspect(other)}"}
  end
end
```

Feel free to use whatever parts you want (untested).

Reply:
Thank you so much for the refactoring! We're in a bit of a rush to deliver this feature, so I apologize for the messy code 😖